### PR TITLE
added a sample policy for readwriteoncepod

### LINKED ISSUE #853: [Sample] policy to enforce single pod access mode for persistent volumes
### Problem Statement

See: https://kubernetes.io/blog/2023/04/20/read-write-once-pod-access-mode-beta/

### Solution Description

Add sample policy to enforce `ReadWriteOncePod` access mode.

### Example "Good" Resource

_No response_

### Example "Bad" Resource

_No response_

### Other Comments

_No response_

### Slack discussion

_No response_

### Troubleshooting

- [X] I have read and followed the documentation AND the [troubleshooting guide](https://kyverno.io/docs/troubleshooting/).
- [X] I have verified the policy does not exist in the [samples library](https://kyverno.io/policies/).

### PR DESCRIPTION
**Pull request by swastik959**
## Related Issue(s)
closes #853

## Description
added a sample policy for readwriteoncepod

## Checklist

- [] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
- [] I have added the artifacthub-pkg.yml file and have verified it is complete and correct.

### CODE REVIEW

**Comment by swastik959** [comment]
@chipzoller please review this pr then I will be adding the tests

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: ReadWriteOnce-Pod
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:4]: lower case names

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  annotations:
+    policies.kyverno.io/title: Enforce ReadWriteOncePod
+    policies.kyverno.io/category: Sample
+    policies.kyverno.io/minversion: 1.6.0
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:8]: Remove

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    policies.kyverno.io/subject: PersistentVolumeClaim
+    policies.kyverno.io/description: >-
+      This validation is valuable when the persistent volumes should be
+      accessed by one single pod it can be useful in an environment where
+      the volumes should be restrictive and should be accessed by only handfull
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:13]:
```suggestion
    policies.kyverno.io/description: >-
      Some stateful workloads with multiple replicas only allow a single Pod to write
      to a given volume at a time. Beginning in Kubernetes 1.22 and enabled by default
      in 1.27, a new setting called ReadWriteOncePod, available
      for CSI volumes only, allows volumes to be writable from only a single Pod. For more
      information see the blog https://kubernetes.io/blog/2023/04/20/read-write-once-pod-access-mode-beta/.
      This policy enforces that the accessModes for a PersistentVolumeClaim be set to ReadWriteOncePod.
```

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+spec:
+  validationFailureAction: audit
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:15]:
```suggestion
  validationFailureAction: Audit
```

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  background: true
+  rules:
+  - name: ReadWriteOnce-Pod
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:18]: lower case rule name, different from policy name

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    match:
+      any:
+      - resources:
+          kinds:
+          - PersistentVolumeClaim
+    validate:
+      message: "access mode of persistent volume claims not set to ReadWriteOncePod"
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:25]:
```suggestion
      message: "The accessMode must be set to ReadWriteOncePod."
```

**Review by chipzoller — CHANGES_REQUESTED** [review]
1. Please use GitHub closing keywords to link PR to issue.
2. Have you tested this policy to ensure it works as intended? I see a problem here.
3. This should be re-written as a standard validate pattern and not a deny condition rule due to its simplicity.
4. Any changes requested in the policy please incorporate into the `artifacthub-pkg.yml` file accordingly.

**Comment by swastik959** [comment]
hi @chipzoller  I tried to run the tests locally but both of them were failing

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: readwriteOnce-pod
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:4]:
```suggestion
  name: readwriteonce-pod
```

Can't have upper case letters in name.

File: other/enforce-readwriteonce-pod/.kyverno-test/kyverno-test.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+kind: Test
+metadata:
+  name: enforce-readwriteonce-pod
+policies:
+- ../enforce-readwriteonce-pod.yml
+resources:
+- resource.yaml
+results:
+- kind: PersistentVolumeClaim
+  policy: enforce-readwriteonce-pod
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/.kyverno-test/kyverno-test.yaml:11]: Name here has to be name of the policy, not the file.

File: other/enforce-readwriteonce-pod/.kyverno-test/kyverno-test.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  resources:
+  - badpvc
+  result: fail
+  rule: readwrite-pvc-single-pod
+- kind: PersistentVolumeClaim
+  policy: enforce-readwriteonce-pod
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/.kyverno-test/kyverno-test.yaml:17]: Ditto

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: other/enforce-readwriteonce-pod/artifacthub-pkg.yml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+description: >-
+      Some stateful workloads with multiple replicas only allow a single Pod to write
+      to a given volume at a time. Beginning in Kubernetes 1.22 and enabled by default
+      in 1.27, a new setting called ReadWriteOncePod, available
+      for CSI volumes only, allows volumes to be writable from only a single Pod. For more
+      information see the blog https://kubernetes.io/blog/2023/04/20/read-write-once-pod-access-mode-beta/.
+      This policy enforces that the accessModes for a PersistentVolumeClaim be set to ReadWriteOncePod.
+install: |-
+  ```shell
+  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yaml
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/artifacthub-pkg.yml:14]:
```suggestion
  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml
```

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml:1]: Rename to `.yaml`

File: other/enforce-readwriteonce-pod/artifacthub-pkg.yml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+description: >-
+      Some stateful workloads with multiple replicas only allow a single Pod to write
+      to a given volume at a time. Beginning in Kubernetes 1.22 and enabled by default
+      in 1.27, a new setting called ReadWriteOncePod, available
+      for CSI volumes only, allows volumes to be writable from only a single Pod. For more
+      information see the blog https://kubernetes.io/blog/2023/04/20/read-write-once-pod-access-mode-beta/.
+      This policy enforces that the accessModes for a PersistentVolumeClaim be set to ReadWriteOncePod.
+install: |-
+  ```shell
+  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yml
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/artifacthub-pkg.yml:14]:
```suggestion
  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yaml
```

File: other/enforce-readwriteonce-pod/.kyverno-test/kyverno-test.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
+apiVersion: cli.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: enforce-readwriteonce-pod
+policies:
+- ../enforce-readwriteonce-pod.yml
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/.kyverno-test/kyverno-test.yaml:6]:
```suggestion
- ../enforce-readwriteonce-pod.yaml
```

File: other/enforce-readwriteonce-pod/artifacthub-pkg.yml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+      Some stateful workloads with multiple replicas only allow a single Pod to write
+      to a given volume at a time. Beginning in Kubernetes 1.22 and enabled by default
+      in 1.27, a new setting called ReadWriteOncePod, available
+      for CSI volumes only, allows volumes to be writable from only a single Pod. For more
+      information see the blog https://kubernetes.io/blog/2023/04/20/read-write-once-pod-access-mode-beta/.
+      This policy enforces that the accessModes for a PersistentVolumeClaim be set to ReadWriteOncePod.
+annotations:
+  kyverno/category: "Sample"
+  kyverno/subject: "PersistentVolumeClaims"
+digest: 408c8011fc0befd3648fc5b89f14e9ac9307e836d2feaa1389ed52bc315643cb
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/artifacthub-pkg.yml:29]: Digest is not correct and there is no new line at the end of this file.

File: other/enforce-readwriteonce-pod/artifacthub-pkg.yml [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+      information see the blog https://kubernetes.io/blog/2023/04/20/read-write-once-pod-access-mode-beta/.
+      This policy enforces that the accessModes for a PersistentVolumeClaim be set to ReadWriteOncePod.
+install: |-
+  ```shell
+  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yaml
+  ```
+keywords:
+  - kyverno
+  - Sample
+readme: |
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/artifacthub-pkg.yml:19]: Readme is missing the final paragraph common to others. Please look at another sample to copy and paste.

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yaml [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+          kinds:
+          - PersistentVolumeClaim
+    validate:
+      message: "The accessMode must be set to ReadWriteOncePod."
+      pattern:
+        spec:
+          accessModes:
+          - ReadWriteOncePod
+
+
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/enforce-readwriteonce-pod.yaml:33]: Only a single new line is needed. You will need to recompute the hash after any changes to this file.

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Comment by chipzoller** [comment]
The Chainsaw test is failing as you can probably see. Have you tested this policy thoroughly prior to submission? Policies which do not work in all scenarios locally will not pass tests in CI.

File: other/enforce-readwriteonce-pod/.chainsaw-test/chainsaw-step-01-assert-1.yaml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: enforce-readwriteonce-pod
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/.chainsaw-test/chainsaw-step-01-assert-1.yaml:4]: This is not the name of the policy.
```suggestion
  name: readwriteonce-pod
```

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: other/enforce-readwriteonce-pod/.chainsaw-test/chainsaw-step-01-assert-1.yaml (hunk 16 above)
Comment (swastik959) [reply other/enforce-readwriteonce-pod/.chainsaw-test/chainsaw-step-01-assert-1.yaml:4]: === STDERR
        error: resource mapping not found for name: "readwriteonce-pod" namespace: "" from "STDIN": no matches for kind "ClusterPolicy" in version "kyverno.io/v1"
        ensure CRDs are installed first 

this is the error that is coming when I ran the test

Comment (chipzoller) [reply other/enforce-readwriteonce-pod/.chainsaw-test/chainsaw-step-01-assert-1.yaml:4]: Kyverno must be installed in a live cluster before running Chainsaw tests.

File: other/enforce-readwriteonce-pod/artifacthub-pkg.yml [hunk 17]
```diff
@@ -28,4 +28,4 @@ readme: |
 annotations:
   kyverno/category: "Sample"
   kyverno/subject: "PersistentVolumeClaims"
-digest: 681fc0e98d14ae0a099580e221d9a467afa6d8e639b31cd1ac44e2275f7ebf83 
+digest: 2fb28419ac572f7a3d8ee50f6dd157b00b015638f6a9d06728783f35ffb6c7f0
```
Comment (chipzoller) [inline other/enforce-readwriteonce-pod/artifacthub-pkg.yml:31]: Why did you change the digest in this commit? The digest only gets changed when the policy is changed.

**Comment by chipzoller** [comment]
This Chainsaw test is going to fail on Kubernetes versions less than 1.27 because the value isn't recognized. I think what's best is to just remove the Chainsaw test altogether, unfortunately.

File: other/enforce-readwriteonce-pod/artifacthub-pkg.yml (hunk 17 above)
Comment (swastik959) [reply other/enforce-readwriteonce-pod/artifacthub-pkg.yml:31]: actually the ci for linting was failing so I changed it

**Comment by swastik959** [comment]
@chipzoller can you please have a look

**Comment by swastik959** [comment]
@chipzoller I believe there has been been conflicts in the merging due to which ci are failing ??

**Comment by chipzoller** [comment]
It was just a flaky test for some reason.

**Comment by swastik959** [comment]
@chipzoller if there are no further problems can you please merge it

**Review by chipzoller — APPROVED** [review]
Good work.
